### PR TITLE
Think it is fixed, test are still passing, need to make a new test

### PR DESCRIPTION
**Pull request by OGaboriault**
# Description of the problem

- Periodic boundary conditions were not working when they were not declare as the last BC in the parameter file.
- BC_type was changing value at every BC being parsed. "BC_type" was being checked later in the code (in dem.cc) to set "has_periodic_boundaries" to true or false.  If the periodic BC was being declared before the an other BC, "BC_type" was not set to "BoundaryType::periodic", thus "has_periodic_boundaries = false".

# Description of the solution

- We store every BC type in a vector (BC_types). 
- We check each component of that vector, if it's equal to "BoundaryType::periodic", then "has_periodic_boundaries = true"

# How Has This Been Tested?

- Every test is passing.
- I can modify one of our current test to add a dummy BC at the end of a parameter file, so that it check for this kind of problem.

# Comments

- Not gonna lie, it was not hard a bug spot in the code... wish I had spot this mistake earlier in my results....

### CODE REVIEW

File: source/core/parameters_lagrangian.cc [hunk 1]
```diff
@@ -1240,7 +1242,9 @@ namespace Parameters
           point_on_rotation_axis.insert({counter, Point<3>(zero_tensor)});
         }
 
+      // NOTE This first vector should not be initialized this big.
       outlet_boundaries.reserve(DEM_BC_number);
+      bc_types.reserve(DEM_BC_number);
```
Comment (blaisb) [inline source/core/parameters_lagrangian.cc:1247]: Yeah but how big is DEM_BC_number? Here it is reserved and not initialized btw, so this is not really a big deal...

File: source/dem/read_mesh.cc [hunk 2]
```diff
@@ -15,10 +15,12 @@ read_mesh(const Parameters::Mesh   &mesh_parameters,
 
   attach_grid_to_triangulation(triangulation, mesh_parameters);
 
-  if (bc_params.BC_type ==
-      Parameters::Lagrangian::BCDEM::BoundaryType::periodic)
+  // Check if there's periodic boundaries
```
Comment (blaisb) [inline source/dem/read_mesh.cc:18]:
```suggestion
  // Check if there are periodic boundaries
```

File: source/fem-dem/cfd_dem_coupling.cc [hunk 3]
```diff
@@ -1470,10 +1470,16 @@ CFDDEMSolver<dim>::dem_setup_contact_parameters()
   load_balance_step      = false;
 
   // Check if there's periodic boundaries
```
Comment (blaisb) [inline source/fem-dem/cfd_dem_coupling.cc:1472]:
```suggestion
  // Check if there are periodic boundaries
```

**Review by blaisb — APPROVED** [review]
Just tiny comments

**Comment by blaisb** [comment]
@OGaboriault I think this needs a CHANGELOG entry

File: source/core/parameters_lagrangian.cc (hunk 1 above)
Comment (OGaboriault) [reply source/core/parameters_lagrangian.cc:1247]: You are right, DEM_BC_number is really small

File: CHANGELOG.md [hunk 4]
```diff
@@ -3,6 +3,13 @@
 All notable changes to the Lethe project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/).
 
+## [Master] - 2024-05-02
+
+### Fixed
+
+- MAJOR Periodic boundary conditions were not working if they weren't the last boundary condition being declare in the parameter file. Now, every boundary condition work in which even order they are being declared. [#1110](https://github.com/chaos-polymtl/lethe/pull/1110)
```
Comment (acdaigneault) [inline CHANGELOG.md:10]: Specify that this is related the periodic boundaries for DEM only

File: applications_tests/lethe-particles/periodic_boundary_box.prm [hunk 5]
```diff
@@ -93,9 +93,14 @@ end
 #---------------------------------------------------
 
 subsection DEM boundary conditions
-  set number of boundary conditions = 1
+  set number of boundary conditions = 2
 
-  subsection boundary condition 0
+   subsection boundary condition 0
+    # Choices are fixed_wall|outlet|rotational|translational|periodic
```
Comment (acdaigneault) [inline applications_tests/lethe-particles/periodic_boundary_box.prm:99]: What is fixed_wall actually? 
Also, maybe (in another pr) we should have a test also for rotational/translational and PBC (like a rotating drum with PBC on flat walls) and make sure all PB match well.

**Review by acdaigneault — CHANGES_REQUESTED** [review]
This fix is good, sorry about that haha.
Just change the log to specify that the bug was only related to PBC in DEM since there are PBC for fluid that were not affected by this bug.

**Review by acdaigneault — APPROVED** [review]

**Review by acdaigneault — COMMENTED** [review]
Btw, change the name of the PR haha
